### PR TITLE
Validate `memory_limit` option

### LINKED ISSUE #54: `memory_limit` option is not properly validate when passed via `MemoryOptions::__construct` or set via `MemoryOptions#setMemoryLimit`
Hi,

I'm debugging why my given memory limit is not being applied in the adapter. Looks like there is a bug in the MemoryOptions. I think `preg_match` in this line

https://github.com/laminas/laminas-cache-storage-adapter-memory/blob/d2c357a8b839ceb0e0781d5e9aebe46642dbf0b2/src/MemoryOptions.php#L88

should use the parameter `$value` instead of `ini_get('memory_limit')`?

BR,
Alex

### PR DESCRIPTION
**Pull request by leuchtdiode**
|    Q          |   A
|-------------- | ------
| Bugfix        | yes
| BC Break      | no

### Description

Passing invalid options via `MemoryOptions#setMemoryLimit` or via `MemoryOptions#__construct`, the actual `memory_limit` option is not properly validated.

### CODE REVIEW

File: src/MemoryOptions.php
Comment (Ocramius) [inline src/MemoryOptions.php]: If you didn't recormat this file on purpose,I suggest:

 * changing just what you intended
 * reverting all the rest
 * squashing
 * force pushing

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/MemoryOptions.php
Comment (Ocramius) [inline src/MemoryOptions.php]: Shouldn't this default to the memory limit?

Comment (leuchtdiode) [reply src/MemoryOptions.php]: It does default to the "memory limit divded by 2" in `getMemoryLimit` if `$this->memoryLimit` is `null`. I just think that passing `null` to the `normalizeMemoryLimit` function should return `null`, as this is a value which "can't be normalized".

File: src/MemoryOptions.php
Comment (Xerkus) [inline src/MemoryOptions.php]: I believe this never accepted null. It would always fail on regex failing to match

Comment (leuchtdiode) [reply src/MemoryOptions.php]: When I remove the line, the existing test `testOptionsFluentInterface` is failing, as preg_match does not accept `null` but `string` as argument. It appears this test is trying to set `null` as the current value is `null`. The test looks if setting the current value leads to the same result which is already set.
If I change to `$value ?? ''`, an InvalidArgumentException is thrown (which is correct), but the test still fails as it doesn't expect an exception.

```
1) LaminasTest\Cache\Storage\Adapter\MemoryOptionsTest::testOptionsFluentInterface
TypeError: preg_match(): Argument #2 ($subject) must be of type string, null given

/var/www/html/library/laminas-cache-storage-adapter-memory/src/MemoryOptions.php:92
/var/www/html/library/laminas-cache-storage-adapter-memory/src/MemoryOptions.php:41
/var/www/html/library/laminas-cache-storage-adapter-memory/vendor/laminas/laminas-cache-storage-adapter-test/src/AbstractCommonAdapterTest.php:160
```

Comment (Ocramius) [reply src/MemoryOptions.php]: I think we should cast the value to `(string)`, then let `Exception\InvalidArgumentException` be thrown (below)

Comment (leuchtdiode) [reply src/MemoryOptions.php]: But how to fix the test from the common base test class?

Comment (Ocramius) [reply src/MemoryOptions.php]: It's an invalid option, no? If there is a way to set it to invalid, we'd reach this scenario.

The best course of action would be to completely disallow it being an invalid value in first place :)

Comment (Xerkus) [reply src/MemoryOptions.php]: type error from null passed to preg_match is new in php 8 where native types were added to built-in functions.

The type error is encountered due to missing argument type validation that would have produced different error a line or two earlier. 
Proper approach would be to check if value is not an int an not a string then throw invalid argument exception.

Comment (leuchtdiode) [reply src/MemoryOptions.php]: If I add this instead of null check: 

```
if(! is_string($value) && ! is_int($value)) {
    throw new Exception\InvalidArgumentException('Memory limit must be an integer or string');
}
```

psalm is crying about:

```
ERROR: DocblockTypeContradiction - src/MemoryOptions.php:84:12 - Docblock-defined type int for $value is always int (see https://psalm.dev/155)
        if(! is_string($value) && ! is_int($value)) {
```

Furthermore it still breaks the test `testOptionsFluentInterface` as the test is not expecting the new Exception `Memory limit must be an integer or string`.

File: src/MemoryOptions.php
Comment (boesing) [inline src/MemoryOptions.php:49]:
```suggestion
     * @param string|int $memoryLimit String values have to fulfill the php.ini format while the integer value represents the amount of allowed memory in bytes
```

Or something like this, maybe @froschdesign has a better phrase tho.

File: test/unit/MemoryOptionsTest.php
Comment (boesing) [inline test/unit/MemoryOptionsTest.php:1]: can we have native return types in this class?

File: test/unit/MemoryOptionsTest.php
Comment (boesing) [inline test/unit/MemoryOptionsTest.php]: This is testing the options, IMO no need to test the whole adapter.
You can use `AbstractAdapterOptionsTest` instead.

**Comment by boesing** [comment]
I am not really concerned about this change. This introduces handling for invalid `memory_limit` values. 
I tested this on PHP 8.3 and even PHP itself is yelling at me:

```
→ php -dmemory_limit=M256 -i | grep memory_limit
Using php v8.3.2
PHP Warning:  Invalid "memory_limit" setting. Invalid quantity "M256": no valid leading digits, interpreting as "0" for backwards compatibility in Unknown on line 0
PHP Warning:  Failed to set memory limit to 0 bytes (Current memory usage is 2097152 bytes) in Unknown on line 0
Warning: Invalid "memory_limit" setting. Invalid quantity "M256": no valid leading digits, interpreting as "0" for backwards compatibility in Unknown on line 0
memory_limit => 128M => 128M
```

I would instead enhance the documentation to outline that in case that a string value, passed as `memory_limit` option to the adapter, should properly support PHPs internal memory declaration structure.

Do we have other opinions here? I mean, there was already quite some discussion.

File: test/unit/MemoryOptionsTest.php
Comment (leuchtdiode) [reply test/unit/MemoryOptionsTest.php:1]: Yes I can add them

File: test/unit/MemoryOptionsTest.php
Comment (leuchtdiode) [reply test/unit/MemoryOptionsTest.php]: I changed it, which fixed my test. But the already existing MemoryTest is failing with the same error:

```
1) LaminasTest\Cache\Storage\Adapter\MemoryTest::testOptionsFluentInterface
Laminas\Cache\Exception\InvalidArgumentException: Memory limit must be an integer or string
```

Actually I would leave the null check and return back `null` when `null` is given to `normalizeMemoryLimit`. Why does it hurt?

**Comment by leuchtdiode** [comment]
> I am not really concerned about this change. This introduces handling for invalid `memory_limit` values. I tested this on PHP 8.3 and even PHP itself is yelling at me:
> 
> ```
> → php -dmemory_limit=M256 -i | grep memory_limit
> Using php v8.3.2
> PHP Warning:  Invalid "memory_limit" setting. Invalid quantity "M256": no valid leading digits, interpreting as "0" for backwards compatibility in Unknown on line 0
> PHP Warning:  Failed to set memory limit to 0 bytes (Current memory usage is 2097152 bytes) in Unknown on line 0
> Warning: Invalid "memory_limit" setting. Invalid quantity "M256": no valid leading digits, interpreting as "0" for backwards compatibility in Unknown on line 0
> memory_limit => 128M => 128M
> ```
> 
> I would instead enhance the documentation to outline that in case that a string value, passed as `memory_limit` option to the adapter, should properly support PHPs internal memory declaration structure.
> 
> Do we have other opinions here? I mean, there was already quite some discussion.

I'm with you at this. Why would someone expect a format to work which is no standard anywhere?

**Comment by boesing** [comment]
> I'm with you at this. Why would someone expect a format to work which is no standard anywhere?

But that is what you are trying to support with this PR, no?
At least from what I can see in the test, you want to allow `M256` as a valid memory limit. If that is not what you try to achieve, why are these changes needed in the first place? 😅

**Comment by leuchtdiode** [comment]
> > I'm with you at this. Why would someone expect a format to work which is no standard anywhere?
> 
> But that is what you are trying to support with this PR, no? At least from what I can see in the test, you want to allow `M256` as a valid memory limit. If that is not what you try to achieve, why are these changes needed in the first place? 😅

No, this is a misunderstanding. The only thing I wanted to fix is the usage of `ini_get('memory_limit')` instead of the given parameter `$value` in `normalizeMemoryLimit`. This PR just got out of control because the tests are failing now because of my change 😂 And it raised the question on my side if we should prevent these wrong values from coming in.

I just added the test to "show" that it is currently wrong because it does return 256 (bytes). M256 and G1024 would go out of the test when we found a solution. Sorry, this was confusing from my side to put the two values into the data set.

**Comment by boesing** [comment]
Okay, let me rephrase what we want to achieve with this PR:
- we do want to fix `normalizeMemoryLimit` method so that the value passed to `setMemoryLimit` is validated instead of the value persisted in `memory_limit` ini setting
- we do want to have a **failing test** for those memory sizes you accidentally added to the valid ones, i.e. `M256` and `G1024`
- we do want to keep some of the other unit tests which are actually quite nice to have, I'd love to keep these

Did I forgot something? Let me know!

Could you probably refactor this PR to have only those changes in place? Happy to merge that once thats done (maybe not today but definitely in the next days). That would be perfect. Thanks in advance for contributing this!


----

> I just added the test to "show" that it is currently wrong because it does return 256 (bytes). M256 and G1024 would go out of the test when we found a solution. Sorry, this was confusing from my side to put the two values into the data set.

Ah I see, well that escalated quickly 😂  However, lets try to focus on the initial problem. I'd just add a proper failing test for the invalid values. That should outline that the memory limit is not validated as it should be. Then we can have the fix in place and are good to go.

**Comment by leuchtdiode** [comment]
> Okay, let me rephrase what we want to achieve with this PR:
> 
> * we do want to fix `normalizeMemoryLimit` method so that the value passed to `setMemoryLimit` is validated instead of the value persisted in `memory_limit` ini setting
> * we do want to have a **failing test** for those memory sizes you accidentally added to the valid ones, i.e. `M256` and `G1024`
> * we do want to keep some of the other unit tests which are actually quite nice to have, I'd love to keep these
> 
> Did I forgot something? Let me know!
> 
> Could you probably refactor this PR to have only those changes in place? Happy to merge that once thats done (maybe not today but definitely in the next days). That would be perfect. Thanks in advance for contributing this!

Yes, that's correctly summarized.

Point 1: Already done.
Point 2: I will change
Point 3: Does anyone have a problem with the null check in `normalizeMemoryLimit`  This would let the existing tests pass.

File: test/unit/MemoryOptionsTest.php
Comment (boesing) [reply test/unit/MemoryOptionsTest.php]: Can you point me to the test which tests `null` to be a valid value? 
That sounds kinda weird to me and I wonder how we are ending up in that error.
IMO, I'd disallow `null` at all but maybe we have something wrong in our test which leads to that problem.

File: src/MemoryOptions.php
Comment (boesing) [reply src/MemoryOptions.php]: I'd say lets not allow `null` here and instead do the memory_limit divided by 2 thing in the constructor. Would that work? (Havent had a look into the current implementation, just a wild guess).
That would also give us a meaningful property default value.

File: test/unit/MemoryOptionsTest.php
Comment (leuchtdiode) [reply test/unit/MemoryOptionsTest.php]: It is this already existing test:

```
1) LaminasTest\Cache\Storage\Adapter\MemoryTest::testOptionsFluentInterface
Laminas\Cache\Exception\InvalidArgumentException: Memory limit must be an integer or string
```

Looks like this test is just setting the current value of the options property and expects the same value back again. As there is nothing set for `memoryLimit`by default, the test throws an exception now instead of returning `null` which breaks the test.

**Comment by leuchtdiode** [comment]
I adapted the things we discussed. Please have a look if that suits you. Maybe you can double check the regex which I had to adapt to disallow M128, G1024, etc.
That i hopefully did not oversee something :)

The code formatting makes me crazy. I have to check how to apply phpcs.xml in PHPStorm 😂

**Comment by boesing** [comment]
> Does anyone have a problem with the null check in normalizeMemoryLimit This would let the existing tests pass.

🤚🏼 
Can you point me to the test which is failing so that I can have a look into that one before merging?

File: src/MemoryOptions.php
Comment (boesing) [inline src/MemoryOptions.php]: I think I'm fine with this one. I would say lets remove the `$` for now.
I still wonder why a negative numeric value would be okay here but who am I to judge this ancient code.
Most probably only @weierophinney or @marc-mabe are aware of that and thus lets keep it for now. Could be removed in next major.
I most probably keep that in mind until then.

File: test/unit/MemoryOptionsTest.php
Comment (boesing) [reply test/unit/MemoryOptionsTest.php]: It is fixed now due to the default value for the `memoryLimit`, no? The only reason why `null` would be passed which I could imagine (and I was most probably right) was that the adapter test somehow uses `AdapterOptions#toArray` (or sth. similar) to then pass all those key/value patterns to the setter methods.
Due to the lack of a default value for memory size, `null` was used for the setter which ended up in the failing test.

**Review by boesing — CHANGES_REQUESTED** [review]
I think after dropping the `$` in the regex, this PR is ready to get merged.

File: test/unit/MemoryOptionsTest.php
Comment (leuchtdiode) [reply test/unit/MemoryOptionsTest.php]: Yes, that is exactly the case. Setting the default is letting the test pass.

**Comment by leuchtdiode** [comment]
> I think after dropping the `$` in the regex, this PR is ready to get merged.

Done :)

**Comment by Xerkus** [comment]
Why `$` is dropped from regex? it should match full string
oof. Regex had no full match constraints? it could match weirdest stuff... so much for input validation

File: src/MemoryOptions.php
Comment (Xerkus) [reply src/MemoryOptions.php]: negative numeric is for -1 = no limit

Comment (boesing) [reply src/MemoryOptions.php]: Yup, but thats how it worked before as well.
If s1 wants to set unlimited memory, thats fine in the current major as thats how it works since like forever.
I am open to discuss this in an RFC for the next major version, would you agree on that?

**Comment by boesing** [comment]
> Why `$` is dropped from regex? it should match full string oof. Regex had no full match constraints? it could match weirdest stuff... so much for input validation

Because it was introduced in this patch. Passing something like `1000 K YALLA` is parsed as `1000` and `K`. I am not even super happy with the `^` because you could have passed `YALLA 1000 K YALLA` and it worked. But I do not really think that this might introduce too much noise but something like `1000 K ` (see that whitespace at the end?) should still be parsable (for now).

I think in next major, we should do something like:
```php
if (is_numeric($value)) {
    return (int) $value;
}

if (! preg_match('/^(?<value>\-?\d+)\s+?(?<unit>[GMK])?$/i', trim($value), $matches)) {
    throw new Exception\InvalidArgumentException(sprintf(
        'Invalid memory limit "%s" passed. Must be compatible with PHP shorthand bytes notation',
        $value,
    ));
}

$value (int) $matches['value'];
$unit = strtoupper($matches['unit']);

if ($value < -1) {
    throw new Exception\InvalidArgumentException('Provided memory limit must represent a value greater than or equal to `-1`');
}

return match ($unit) {
   'G' => $value * 1024 * 1024 * 1024,
   'M' => $value * 1024 * 1024,
   'K' => $value * 1024,
   default => $value,
};
```

Fine for every1?

**Comment by leuchtdiode** [comment]
> > Why `$` is dropped from regex? it should match full string oof. Regex had no full match constraints? it could match weirdest stuff... so much for input validation
> 
> Because it was introduced in this patch. Passing something like `1000 K YALLA` is parsed as `1000` and `K`. I am not even super happy with the `^` you could have passed `YALLA 1000 K YALLA` and it worked. But I do not really think that this might introduce too much noise but something like `1000 K ` (see that whitespace at the end?) should still be parsable.
> 
> I think in next major, we should do something like:
> 
> ```
> if (is_numeric($value)) {
>     return (int) $value;
> }
> 
> if (! preg_match('/^(?<value>\-?\d+)\s+?(?<unit>[GMK])?$/i', trim($value), $matches)) {
>     throw new Exception\InvalidArgumentException(sprintf(
>         'Invalid memory limit "%s" passed. Must be compatible with PHP shorthand bytes notation',
>         $value,
>     ));
> }
> 
> $value (int) $matches['value'];
> $unit = strtoupper($matches['unit']);
> 
> if ($value < -1) {
>     throw new Exception\InvalidArgumentException('Provided memory limit must represent a value greater than or equal to `-1`');
> }
> 
> return match ($unit) {
>    'G' => $value * 1024 * 1024 * 1024,
>    'M' => $value * 1024 * 1024,
>    'K' => $value * 1024,
>    default => $value,
> };
> ```
> 
> Fine for every1?

Sounds good to me. I will add "-1" and -1 to the valid data set.

**Comment by boesing** [comment]
FYI: I've created an RFC (#57) regarding the memory limit to have it removed from v3.0.0
I'd love to get some feedback there, esp. since symfony `ArrayAdapter` (which is the `Memory` equivalent of symfony framework) does not have this kind of feature. Rationale is stated within the RFC.

**Pull request closed without merging**
